Sources/TSCBasic/FileSystem.swift

```diff
@@ -12,74 +12,87 @@ import TSCLibc
 import Dispatch
 
-public enum FileSystemError: Swift.Error {
-    /// Access to the path is denied.
-    ///
-    /// This is used when an operation cannot be completed because a component of
-    /// the path cannot be accessed.
-    ///
-    /// Used in situations that correspond to the POSIX EACCES error code.
-    case invalidAccess
-
-    /// IO Error encoding
-    ///
-    /// This is used when an operation cannot be completed due to an otherwise
-    /// unspecified IO error.
-    case ioError
-
-    /// Is a directory
-    ///
-    /// This is used when an operation cannot be completed because a component
-    /// of the path which was expected to be a file was not.
-    ///
-    /// Used in situations that correspond to the POSIX EISDIR error code.
-    case isDirectory
-
-    /// No such path exists.
-    ///
-    /// This is used when a path specified does not exist, but it was expected
-    /// to.
-    ///
-    /// Used in situations that correspond to the POSIX ENOENT error code.
-    case noEntry
-
-    /// Not a directory
-    ///
-    /// This is used when an operation cannot be completed because a component
-    /// of the path which was expected to be a directory was not.
-    ///
-    /// Used in situations that correspond to the POSIX ENOTDIR error code.
-    case notDirectory
-
-    /// Unsupported operation
-    ///
-    /// This is used when an operation is not supported by the concrete file
-    /// system implementation.
-    case unsupported
-
-    /// An unspecific operating system error.
-    case unknownOSError
-
-    /// File or folder already exists at destination.
-    ///
-    /// This is thrown when copying or moving a file or directory but the destination
-    /// path already contains a file or folder.
-    case alreadyExistsAtDestination
+public struct FileSystemError: Swift.Error, Equatable {
+    public enum Kind: Equatable {
+        /// Access to the path is denied.
+        ///
+        /// This is used when an operation cannot be completed because a component of
+        /// the path cannot be accessed.
+        ///
+        /// Used in situations that correspond to the POSIX EACCES error code.
+        case invalidAccess
+
+        /// IO Error encoding
+        ///
+        /// This is used when an operation cannot be completed due to an otherwise
+        /// unspecified IO error.
+        case ioError(code: Int32)
+
+        /// Is a directory
+        ///
+        /// This is used when an operation cannot be completed because a component
+        /// of the path which was expected to be a file was not.
+        ///
+        /// Used in situations that correspond to the POSIX EISDIR error code.
+        case isDirectory
+
+        /// No such path exists.
+        ///
+        /// This is used when a path specified does not exist, but it was expected
+        /// to.
+        ///
+        /// Used in situations that correspond to the POSIX ENOENT error code.
+        case noEntry
```

Comment:
Do we want to record any information here about whether symlink traversal was in effect when the error occurred, e.g. if I'm accessing path where a symlink is the last path component, I would get this error if a symlink was being resolved but not get it if it wasn't.  This is known at the callsite but cannot be inferred from the error.  I suppose that is actually something that could apply to any of the errors here.

Reply:
How would you prefer to record that? A separate property on this new `struct`?

Reply:
I'm not sure, actually — at first I was thinking a boolean on each of the enum cases where it made sense, but perhaps it would belong as a separate boolean next to the path.  But this could easily be added later in a separate PR, I think.